demo-project/demo-frontend/login.js

```diff
@@ -1,8 +1,29 @@
-// add click listener to button to trigger login
-const loginButton = document.getElementById('login');
-loginButton.addEventListener('click', () => {
-    const clientId = document.getElementById('client-id').value;
-    const nonce = document.cookie.split('; ').find(row => row.startsWith('nonce')).split('=')[1];
-    const state = document.cookie.split('; ').find(row => row.startsWith('state')).split('=')[1];
-    window.location.href = `http://localhost:3001/api/v1/authorize?response_type=id_token&client_id=${clientId}&redirect_uri=http://localhost:1606/callback.html&scope=openid&state=${state}&nonce=${nonce}`;
+// add click listeners to login buttons.
+const loginButtonImplicit = document.getElementById('login-implicit');
+loginButtonImplicit.addEventListener('click', () => {
+  const clientId = getClientIdElement();
+  const nonce = getNonce();
+  const state = getState();
+  window.location.href = `http://localhost:3001/api/v1/authorize?response_type=id_token&client_id=${clientId}&redirect_uri=http://localhost:1606/callback.html&scope=openid&state=${state}&nonce=${nonce}`;
 });
+
+const loginButtonAuthCode = document.getElementById('login-authorization-code');
+loginButtonAuthCode.addEventListener('click', () => {
+  const clientId = getClientIdElement();
+  const nonce = getNonce();
+  const state = getState();
+  window.location.href = `http://localhost:3001/api/v1/authorize?response_type=code&client_id=${clientId}&redirect_uri=http://localhost:1606/callback.html&scope=openid&state=${state}&nonce=${nonce}`;
```

Comment:
The difference between these two buttons is only the `window.location.href` property and the ID selector, right? Can we refactor this so that we only have a button? THe rest of the behaviour seems to be exactly the same (and it should be).

Reply:
No I mean refactor the code that generates the button, so that both can be created with the same function, pulling out the stuff that is different in an argument.

Reply:
abfe3f3